### PR TITLE
Central node with Relay behaviour

### PR DESCRIPTION
**Pull request by MitaliBo**
## Description

Fixes pyrsia/pyrsia#

This PR introduces relay protocol for central/authoritative node and also separation of network code from main node. This is the first setup for hole punching is to have relay node set.
## PR Checklist

- [x] I've read the [contributing guidelines](https://github.com/pyrsia/.github/blob/main/contributing.md).
- [x] I've read ["What is a Good PR?"](https://github.com/pyrsia/pyrsia/blob/main/docs/good_pr.md)
- [x] I've included a good title and brief description along with how to review them.
- [x] I've linked any associated an [issue](https://github.com/pyrsia/pyrsia/issues).
- [x] I've requested a review  from `pyrsia/collaborators`.

### Code Contributions

- [ ] I've built the code `cargo build --all-targets` successfully.
- [ ] I've run the unit tests `cargo test --workspace` and everything passes.
- [ ] I've made sure my rust toolchain is current `rustup update`.

### CODE REVIEW

**Review by prince-chrismc — DISMISSED** [review]
Do we really need a separate exe for something as simply as adding a relay to the behavior?
There's a tremendous development effort in duplicating the 1.5k lines of diff. I am not sure it's worth it. 

What is the consequence of every node having this behavior?

We should have a meeting to decide this.

**Comment by fishseabowl** [comment]
The central node has too much workload. Maybe a standalone relay node is better.

**Comment by MitaliBo** [comment]
> Do we really need a separate exe for something as simply as adding a relay to the behavior? There's a tremendous development effort in duplicating the 1.5k lines of diff. I am not sure it's worth it.
> 
> What is the consequence of every node having this behavior?
> 
> We should have a meeting to decide this.

@prince-chrismc This is just the first PR for separating out code, server will have more protocols like autonat and other nodes will have dcutr which will again have listen and dial mode, if we want to keep everything at one place that means specifying a whole lot of protocols that we want to enable for a node with command line. I don't see"tremendous effort" to have different binaries for them. Also we want to have all other nodes bootstrapped with central node. I did specify this multiple times during standup but sure we can have a meeting to discuss.

**Comment by prince-chrismc** [comment]
>This is just the first PR for separating out code, server will have more protocols like autonat and other nodes will have dcutr which will again have listen and dial mode

It was not clear that there needs to be specific "client" and "server" implementations.

Can you please share a diagram of what protocols need to be implemented in which different version? This sounds specific to the libp2p rust implementation so it might be worth drawing that.

**Comment by MitaliBo** [comment]
> > This is just the first PR for separating out code, server will have more protocols like autonat and other nodes will have dcutr which will again have listen and dial mode
> 
> It was not clear that there needs to be specific "client" and "server" implementations.
> 
> Can you please share a diagram of what protocols need to be implemented in which different version? This sounds specific to the libp2p rust implementation so it might be worth drawing that.

This is here 
https://docs.google.com/presentation/d/1LpFb_H2njL9iR8fIK1eWUo3YhVPtcCSgWE-Fcse1aqQ/edit?usp=sharing

**Comment by prince-chrismc** [comment]
Since there's a login screen for the public... let me try to summarize

There's three protocols:
1. AutoNAT
2. Relay
3. [DCUtR](https://github.com/libp2p/specs/blob/master/relay/DCUtR.md) 🧙‍♀️ 

```mermaid
sequenceDiagram
    Node_0->>Node_1: Auto NAT detection test
    Node_0->>Central_Node: Open Relay to Node 1
    Central_Node->>Node_1: DCUtR open connection
    Node_0-->>Node_1: Hello!
```

File: src/network_central/p2p.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            create_transport(keypair)?,
+            PyrsiaNetworkBehaviour {
+                identify: identify::Identify::new(identify_config),
+                relay: Relay::new(
+                    peer_id,
+                    relay::Config {
+                        reservation_duration: Duration::from_secs(86400), // 1 hour
+                        ..Default::default()
+                    },
+                ),
```
Comment (fishseabowl) [inline src/network_central/p2p.rs:163]: Is it a relay configuration?

File: src/network_central/event_loop.rs [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+            relay::Event::CircuitReqDenied { .. } => {}
+            relay::Event::CircuitReqDenyFailed { .. } => {}
+            relay::Event::CircuitReqAccepted { .. } => {}
+            relay::Event::CircuitReqOutboundConnectFailed { .. } => {}
+            relay::Event::CircuitReqAcceptFailed { .. } => {}
+            relay::Event::CircuitClosed { .. } => {}
+
+            other => debug!("Unexpected relay behaviour event: {:?}.", other),
+        }
+    }
```
Comment (fishseabowl) [inline src/network_central/event_loop.rs:257]: It is better to add todo macro. 

```rust
todo!()
```

File: src/network_central/p2p.rs (hunk 1 above)
Comment (MitaliBo) [reply src/network_central/p2p.rs:163]: yes

**Comment by MitaliBo** [comment]
> Since there's a login screen for the public... let me try to summarize
> 
> There's three protocols:
> 
> 1. AutoNAT
> 2. Relay
> 3. [DCUtR](https://github.com/libp2p/specs/blob/master/relay/DCUtR.md) 🧙‍♀️

This is incorrect flow, please follow the correct understanding of flow from here: https://docs.google.com/presentation/d/1LpFb_H2njL9iR8fIK1eWUo3YhVPtcCSgWE-Fcse1aqQ/edit#slide=id.p

**Review by fishseabowl — APPROVED** [review]
LGTM

**Review by chb0github — DISMISSED** [review]

**Comment by prince-chrismc** [comment]
This was discussed on May 3rd's standups, we are going to wait for #574 so there are no more conflicts (priority ordering)

**Review by fishseabowl — APPROVED** [review]

**Comment by MitaliBo** [comment]
Merged all the upstream changes, requesting re-review

**Review by johanvos — APPROVED** [review]
Looks good, good basis for the follow-up work.

**Review by prince-chrismc — APPROVED** [review]

**Comment by MitaliBo** [comment]
> Quick general comment: I see new files with (c) 2021 -- best to start with 2022, and while editting files, it is probably good practice to check the header of that file and modify it (in case changes are made to that file)

I was being told that we will handle the upgrade of license year in separate PR as changing as of now will break the CI.

**Comment by prince-chrismc** [comment]
> I was being told that we will handle the upgrade of license year in separate PR as changing as of now will break the CI.

https://github.com/pyrsia/pyrsia/blob/cd8ccb581232fc28512198e8f72ba5bf49e18c9d/.github/license-check/header-Apache-2.0.txt#L2

We are still waiting to see what will come from OpenSSF so we will fix everything at once 😄

**Pull request merged**